src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java

```diff
@@ -363,54 +363,144 @@ private String breadcrumbs() {
-            SEQUENCE_ELEMENT_INDEX("bound sequence element"),
-            SEQUENCE_RANGE("sequence range"),
-            GROUP_ELEMENT("group element"),
-            DEREF_ELEMENT("dereference element");
+        // Assert invariants
+        public GroupElementByName {
+            Objects.requireNonNull(name);
+        }
+
+        @Override
+        public LayoutPath apply(LayoutPath layoutPath) {
+            return layoutPath.groupElement(name);
+        }
+
+        @Override
+        public String toString() {
+            return "groupElement(\"" + name + "\")";
+        }
+    }
+
+    public record GroupElementByIndex(long index)
+            implements MemoryLayout.PathElement, UnaryOperator<LayoutPath> {
+
+        // Assert invariants
+        public GroupElementByIndex {
+            if (index < 0) {
+                throw new IllegalArgumentException("Index < 0");
+            }
+        }
+
+        @Override
+        public LayoutPath apply(LayoutPath layoutPath) {
+            return layoutPath.groupElement(index);
+        }
 
-            final String description;
+        @Override
+        public String toString() {
+            return "groupElement(" + index + ")";
+        }
 
-            PathKind(String description) {
-                this.description = description;
+    }
+
+    public record SequenceElementByIndex(long index)
+            implements MemoryLayout.PathElement, UnaryOperator<LayoutPath> {
+
+        // Assert invariants
+        public SequenceElementByIndex {
+            if (index < 0) {
+                throw new IllegalArgumentException("Index < 0");
             }
+        }
+
+        @Override
+        public LayoutPath apply(LayoutPath layoutPath) {
+            return layoutPath.sequenceElement(index);
+        }
+
+        @Override
+        public String toString() {
+            return "sequenceElement(" + index + ")";
+        }
+
+    }
+
+    public record SequenceElementByRange(long start, long step)
+            implements MemoryLayout.PathElement, UnaryOperator<LayoutPath> {
 
-            public String description() {
-                return description;
+        // Assert invariants
+        public SequenceElementByRange {
+            if (start < 0) {
+                throw new IllegalArgumentException("Start index must be positive: " + start);
+            }
+            if (step == 0) {
+                throw new IllegalArgumentException("Step must be != 0: " + step);
             }
         }
 
-        final PathKind kind;
-        final UnaryOperator<LayoutPath> pathOp;
-        final String stringRepresentation;
+        @Override
+        public LayoutPath apply(LayoutPath layoutPath) {
+            return layoutPath.sequenceElement(start, step);
+        }
+
+        @Override
+        public String toString() {
+            return "sequenceElement(" + start + ", " + step + ")";
+        }
+
+    }
+
+    public static final class SequenceElement
```

Comment:
Why are these not empty records?

Reply:
I had that in the beginning but converted them to regular classes as they do not have any record components. But maybe it is better to make them all records for consistency.

Reply:
I think so, even less code :-)